torch/lib/c10d/ProcessGroupNCCL.hpp

```diff
@@ -125,9 +132,78 @@ class ProcessGroupNCCL : public ProcessGroup {
     // to the store.
     std::shared_ptr<Store> store_;
 
+    // Store a Future work associated with WorkNCCL.
+    c10::intrusive_ptr<c10::ivalue::Future> futureWork_;
+
     friend class ProcessGroupNCCL;
   };
 
+  // FutureNCCL is a subclass of ivalue's Future. The goal is to use
+  // this class in getFuture API of WorkNCCL. This Future is mostly a
+  // wrapper to synchronize streams appropriately and it mostly enables
+  // the async programming model of CUDA while trying to adhere to the
+  // Future interface.
+  //
+  // FutureNCCL has a reference to WorkNCCL and NCCL collective's outputs.
+  // Its value is NCCL collective's outputs.
+  struct FutureNCCL : at::ivalue::Future {
+   public:
+    explicit FutureNCCL(
+        std::shared_ptr<ProcessGroupNCCL::WorkNCCL> work,
+        std::vector<at::Tensor>& outputs)
+        : at::ivalue::Future(c10::ListType::create(c10::TensorType::get())),
+          work_(work),
+          outputs_(outputs) {}
+
+    // Simply calls WorkNCCL's wait(). It will return after synchronizing
+    // the correct GPU streams to ensure we can have async CUDA execution
+    // and it does not wait for the entire operation to complete on GPU.
+    // If NCCL_BLOCKING_WAIT is enabled, in that case, it will wait for the
+    // entire operation to complete before returning.
+    void wait() override {
+      work_->wait();
+    }
+
+    // FutureNCCL's value is NCCL collective's outputs and callbacks were
+    // invoked inline by addCallback(), so markCompleted is not needed.
+    void markCompleted(at::IValue value) override {
+      C10_THROW_ERROR(Error, "FutureNCCL::markCompleted is not supported.");
+    }
+
+    // Returns NCCL collective's outputs after WorkNCCL wait() returns.
+    at::IValue value() override {
+      work_->wait();
+      return outputs_;
+    }
+
+    const at::IValue& constValue() override {
+      work_->wait();
+      return outputs_;
+    }
+
+    // Add a callback to FutureNCCL. FutureNCCL invokes the callback inline.
+    // Callbacks return a Future (not FutureNCCL).
+    void addCallback(std::function<void(void)> callback) override {
+      // Run callback inline after synchronizeStreams.
+      callback();
```

Comment:
I might have missed some context but why don't we call `wait()` here (or ensure wait() has been called before this)? If the callback is invoked inline and we have not synchronized streams, then as I understand it the callback could do some work on the default stream using the result of the future, but the result won't actually have the value since the async op on the nccl stream did not complete?

Reply:
My understanding is `callback` will be blocked by `fut.value` when it tries to get the value.

Reply:
> My understanding is that callback has to call fut.value to get the value and it should synchronize the streams.

Is this always true? what if a user adds a callback for profiling reason which does not access the value? Or does our code access the value somewhere?

Reply:
Yeah, if user doesn't call value or wait  inside callback. It won't work as expected. :) 
Should we add a warning about that? Or should revert back to the version where we wait before calling callback() inline, and do `assert(work->isCompleted())` instead of `work->wait()` inside `value` and `constvalue`?
cc'ed @pritamdamania87

Reply:
Let's just add a `wait()` here since in the case of profiling or something this wouldn't work as expected.

Reply:
Makes sense. Then, for efficiency I'm removing `wait` inside `value` and just `assert(work->isCompleted()) before returning value.

Reply:
Makes sense. Then, for efficiency I'm removing `wait` inside `value`. It doesn't make sense to call wait multiple times for the then callback case.

Reply:
Let's keep the wait inside value as well :) Its unlikely, but it is possible that the user calls wait() on one thread and value() on another thread. As a result, it's not guaranteed that the streams have been synchronized.

Reply:
Sure,